File: js/roulette.js
// Roulette bet EV/house-edge. Every standard inside/outside bet (straight,
// split, street, corner, six-line, column, dozen, even-money) pays
// (36 / numbersCovered) - 1 to one, which makes every one of them carry the
// *same* house edge on a given wheel: edge = 1 - 36/pocketCount. This is a
// well-established property of roulette and is why no per-bet-type special
// casing is needed here (unlike the anomalous "top line" 0-00-1-2-3 bet on
// American wheels, which is deliberately not offered as a bet type below).

export const WHEEL_POCKETS = { american: 38, european: 37 };

export const BET_NUMBERS_COVERED = {
  straight: 1,
  split: 2,
  street: 3,
  corner: 4,
  six_line: 6,
  column: 12,
  dozen: 12,
  even_money: 18, // red/black, even/odd, high/low
};

/** Net payout (to 1) for a bet covering `numbersCovered` of the wheel's pockets. */
export function payoutFor(numbersCovered) {
  return 36 / numbersCovered - 1;
}

/**
 * House edge as a fraction (e.g. 0.0526 for American single-zero-adjacent).
 * enPartage: European-only "la partage"/"en prison" rule that returns half
 * the stake on an even-money bet when the ball lands on zero, roughly
 * halving the house edge for those specific bets.
 */
export function houseEdge(wheelType, betType, enPartage = false) {
  const pockets = WHEEL_POCKETS[wheelType];
  let edge = 1 - 36 / pockets;
  if (enPartage && wheelType === "european" && betType === "even_money") {
    edge = edge / 2;
  }
  return edge;
}

/**
 * Full bet analysis: payout odds, house edge, and expected value in
 * currency units for a given flat bet amount.
 */
export function analyzeBet(wheelType, betType, betAmount, enPartage = false) {
  const numbersCovered = BET_NUMBERS_COVERED[betType];
  const payout = payoutFor(numbersCovered);
  const edge = houseEdge(wheelType, betType, enPartage);
  const expectedValue = -edge * betAmount;
  return {
    pocketCount: WHEEL_POCKETS[wheelType],
    numbersCovered,
    payoutToOne: payout,
    houseEdge: edge,
    expectedValue,
  };
}

File: verifier/roulette.py
"""Reference implementation of roulette bet EV/house-edge, mirroring
js/roulette.js.

Every standard inside/outside bet pays (36 / numbers_covered) - 1 to one,
which makes every one of them carry the *same* house edge on a given wheel:
edge = 1 - 36/pocket_count. This is a well-established property of roulette;
the anomalous "top line" 0-00-1-2-3 bet on American wheels is deliberately
not offered as a bet type.
"""
from __future__ import annotations

from dataclasses import dataclass

WHEEL_POCKETS = {"american": 38, "european": 37}

BET_NUMBERS_COVERED = {
    "straight": 1,
    "split": 2,
    "street": 3,
    "corner": 4,
    "six_line": 6,
    "column": 12,
    "dozen": 12,
    "even_money": 18,  # red/black, even/odd, high/low
}


def payout_for(numbers_covered: int) -> float:
    return 36 / numbers_covered - 1


def house_edge(wheel_type: str, bet_type: str, en_partage: bool = False) -> float:
    pockets = WHEEL_POCKETS[wheel_type]
    edge = 1 - 36 / pockets
    if en_partage and wheel_type == "european" and bet_type == "even_money":
        edge = edge / 2
    return edge


@dataclass
class BetAnalysis:
    pocket_count: int
    numbers_covered: int
    payout_to_one: float
    house_edge: float
    expected_value: float


def analyze_bet(wheel_type: str, bet_type: str, bet_amount: float, en_partage: bool = False) -> BetAnalysis:
    numbers_covered = BET_NUMBERS_COVERED[bet_type]
    payout = payout_for(numbers_covered)
    edge = house_edge(wheel_type, bet_type, en_partage)
    expected_value = -edge * bet_amount
    return BetAnalysis(
        pocket_count=WHEEL_POCKETS[wheel_type],
        numbers_covered=numbers_covered,
        payout_to_one=payout,
        house_edge=edge,
        expected_value=expected_value,
    )
